### PR TITLE
Swaydb implementation for GlobalSnapshot storage

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (kpudlik) [inline modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:73]: Can we put that to companion object of `SnapshotOrdinal`?

File: modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala
Comment (kpudlik) [inline modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:143]: Alternatively: `tryUpdate(f).ifM(concurrent.unit, loop)`

**Review by kpudlik — CHANGES_REQUESTED** [review]

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:73]: We decided to use order/ordering locally

File: modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala
Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:143]: Implementation is copied from cats

**Review by kpudlik — APPROVED** [review]

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (tbekas) [inline modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:77]: This isn't correctly synchronized. It may lead to storing an invalid chain of snapshots. 

Notation used in example below to represent GlobalSnapshot: `ordinal: thisSnapshotHash -> lastSnapshotHash`

Initial state of the storage [`1: b -> a`]
Fiber 1 wants to save snapshot `2: c -> b`
Fiber 2 wants to save snapshot `2: x -> b`
Fiber 3 wants to save snapshot `3: d -> c`

Flow:
Fiber 1: Reads last snapshot, gets `1: b -> a`
Fiber 2: Reads last snapshot, gets `1: b -> a`
Fiber 1: Validates if the link `2: c -> b` follows `1: b -> a`, it does
Fiber 2: Validates if the link `2: x -> b` follows `1: b -> a`, it does
Fiber 1: Calls `ref(2).update(_ => (2: c -> b))`, storage contains now  [`2: c->b`, `1: b -> a`]
Fiber 3: Reads last snapshot, gets 
Fiber 3: Validates if the link `3: d -> c` follows `2: c -> b`, it does
Fiber 3: Calls `ref(3).update(_ => (3: d -> c))`, storage contains now  [`3: d -> c`, `2: c->b`, `1: b -> a`]
Fiber 2: Calls `ref(2).update(_ => (2: x -> b))`, storage contains now  [`3: d -> c`, `2: x->b`, `1: b -> a`]

The final chain is incorrect `[d, x, b]` while it should be `[d, c, b]`

**Review by tbekas — CHANGES_REQUESTED** [review]

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (tbekas) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:73]: My understanding was that we agreed to use order/ordering locally only if it doesn't include all the fields.

`SnapshotOrdinal` companion object already has an order instance that takes into account all of its fields (which in this case is just a single field).

File: modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala
Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:39]: Yes, there are many constructs to achieve the same thing. I think they are more or less equal, both are equally readable I think

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:95]: I just copied the interface you created, I'll let you change the methods, depending on your needs

File: modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala
Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: Saying that semaphore is an overkill means that there is another, simpler option. Do you know any? I think swaydb can be used only that way to achieve proper concurrent-safety flow. 

I know that we're using mostly lock-free code but that implementation is under the hood. It could be a library that you'll just pull to dependencies.

What's more, we're going to use it for snapshots that are not going to be created thousands per second. I believe it's better to start with such implementation for now. We can think later about finding another storage that can suit us, however this is a good start I think.

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:77]: Ok, I started with another implementation. Now I can make it in a single modify as you did. I'll change that

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:73]: Yeah. SnapshotOrdinal is a newtype for Long so we can move it next to the type, as there will be only one ordering. For other data types I'd use ordering locally until used twice

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (tbekas) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:95]: The previous implementation was using `NonEmptyList` which guaranteed containing at least one element, hence there was just `GlobalSnapshot` returned instead of `Option[GlobalSnapshot]`. You are relaxing the constraints on the underlying data type to `Map` so you reflect these changes in the interface.

Also as a rule of thumb, I think we should review changes independently of what may or may not be introduced in the future.

Comment (marcinwadon) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:95]: I can make it a non empty set if you want to use it like that. I just remember you said it will return Option[GlobalSnapshot] anyway

File: modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala
Comment (tbekas) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: >Saying that semaphore is an overkill means that there is another, simpler option. Do you know any?

An eventually consistent, lock-free and persistent solution for appending values could look like this:

1. Data types:
`SignallingRef[F, Option[(K, V)]]` <- last element only
`MapRef[F, K, Option[V]]` <- last `N` elements cached in memory

2. Subscriptions:
On a stream of changes on `SignalingRef` that adds them them to `MapRef`
On a stream of changes on `SignalingRef` with first `N` elements skipped that removes them from `MapRef`
On a stream of changes on `SignalingRef` that serializes elements and saves them as the local file `tmp/datatype/${key}`

3. Interface
`getLast: F[Option[V]]` - takes the value from `SignalingRef`
`get(key: K): F[Option[V]]` - first checks if the value is present in `MapRef` if not, looks for a local file `tmp/datatype/${key}`
`add(fn: Option[(K, V)] => Option[(K, V)])` - conditionally updates the value in `SignalingRef` (details below)

```
add(fn: Option[(K, V)] => Option[(K, V)]): F[Boolean] = signalingRef.modify { maybeCurrentHead =>
  fn(maybeCurrentHead) match {
    case Some(head) => (head.some, true)
    case None       => (maybeCurrentHead, false)
  }
}
```

File: modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala
Comment (tbekas) [reply modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala:73]: I think that the rule of using order/ordering locally only if it doesn't include all the fields makes more sense than the rule of using it locally only if it's used once. It's hard to keep track of how many orderings for a given type are there in the whole codebase.

File: modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala
Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: So you mention possible solutions but not for the problem you mentioned in the first comment.

Comment (tbekas) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: These problems no longer exist in the solution described above so they don't need to be solved.

Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: That's totally different way of storing data. And semaphore isn't an overkill because overkill means there is something simpler. For swaydb there is no alternative for concurrent-safety.

Comment (tbekas) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: >That's totally different way of storing data.

Exactly. When storing data we need to start from the access patterns that we're going to have. For storing global snapshots we don't need to modify elements by arbitrary key, which is the primary write access pattern supported by SwayDb (a Map interface). Unfortunately, SwayDb doesn't seem to support all the access patterns we need, which are:

write: 
- atomically and conditionally append the head element

read:
- get last element
- get last K elements (possibly in future this will be needed)
- get element by key (ordinal, hash or both, if both then the solution above need to be modified a little bit)

Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: SwayDb supports not only Map but Set as well. It has streaming-api. We can write an ordering that will take N elements only. All read requirements are met in my opinion.

Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala:34]: `ConcurrentHashMap` calculates key hash and uses:
```scala
synchronized (key) {
 ...
}
```
for locking single keys. We can use the same pattern here

**Comment by marcinwadon** [comment]
I'm closing the pull request and gonna work on a different solution for global snapshot storage

**Pull request closed without merging**
